src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/SingleSpaceSeparatorCheck.java

```diff
@@ -199,7 +198,7 @@ private void visitEachToken(DetailAST node) {
                 currentNode = currentNode.getFirstChild();
             }
             else {
-                while (currentNode.getNextSibling() == null && currentNode.getParent() != parent) {
+                while (currentNode.getNextSibling() == null && currentNode.getParent() != null) {
```

Comment:
this will go up to a ROOT ast, we should not do this, as it will walk through all nodes in tree. It is too much.

Reply:
yes, my bad, sorry this is being huge mistake of my. I have not noticed it's a loop

Reply:
@romani it is possible for any situation where the parent will get null rather then  Root. I am not seen any condition for this as of know

Reply:
so I think if we write it as null it will still walk till root ast

Reply:
I recommed to apply mutation as pitest suggest and run diff report on it (do not forget to disable all CI to speedup execution)

Reply:
So you know you don't need CI to run regression.
https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#checkstyle-tester
All this can be run on your local. While we do recommend running against all projects, you can trim the list of projects to get a quick idea if something can be found. It can also let you try different code changes when you do find a difference or such.

Reply:
@Kevin222004 ,
try to remove `currentNode.getParent() != null` completely.

Reply:
> @Kevin222004 , try to remove `currentNode.getParent() != null` completely.

As of know it is becoming quite complicated to me if we remove the `currentNode.getParent() != null` it will working fine and it will traverse the tree as expected. 
but it is giving error when the violation is coming in file. I will still trying to find out the exact reason if you gave any hint. it would be helpfull

Reply:
Just remove it and run diff report generation on your local or by GitHub action.
It is easy.

Reply:
Ok so as a minor update if i remove this   `currentNode.getParent() != null` completely according to https://github.com/checkstyle/checkstyle/pull/12283#discussion_r991607090  so unit test cases are failing in this situation so regression would not be generated there.
 please gave a look at this as well  https://github.com/checkstyle/checkstyle/pull/12283#issuecomment-1272589581

Reply:
@romani ping

Reply:
this this while condition as as.
and apply mutation to `final DetailAST parent = node.getParent();` to make it `final DetailAST parent = node;`
and see what is missed.

sounds like we can remove `final DetailAST parent = node.getParent();` completely and make while condition as
`currentNode.getParent() != node`